### PR TITLE
oop-special_class_method-hard-q13

### PR DESCRIPTION
**Pull request by KaiKhademi**
## PR Checklist
- [x]  Are helper methods really needed here?
    > Opt for hardcoding input/output into the input provider when possible
- [x]  Are the assertion fail messages good?
    - [x]  Logical (does it make sense?)
    - [x]  Readable (can a student understand what’s wrong?)
    - [x]  Not too feed-y (does it give away the answer?)
    - [x]  Grammatically sound (is the grammar and punctuation in the sentence proper?)
    - [x]  Specific (does the fail message relate to the question?)
- [x]  Is the code formatted well?
    > Run `CTRL` + `ALT` + `L` in IntelliJ
- [x]  EOF newlines
- [x]  Do the comments for Parsons follow the correct convention?
- [x]  Is the question being tested sufficiently?
    - [x]  Are there representative cases to test the program?
    - [x]  What about edge cases? (empty strings, negatives, infinity, off-by-one errors, etc)
        > If the question doesn’t specify what to do, mark it as a bad question
- [x]  Are the test method names good?
    > Test method names show up in a “What went well” section on the website, so “mathTest” will become “What went well: math test”, which doesn’t make much sense. A better name in this example would be “calculatesTaxCorrectly” so it shows up as “What went well: calculates tax correctly”

### CODE REVIEW

File: src/oop/special_class_method/hard/q13/QuestionText.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
+# oop-special_class_method-hard-q13
+
+## Question Text
+
+Suppose you were hired by a departmental store to digitize their catalogue system, and one of the classes you are making
+for objects is a **Cup** class. This class has three attributes called **liquid**, **color** , and **volume** which
+cannot be modified, but can be accessed
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q13/QuestionText.md:7]: This says the three attributes have getters but not setters, but everywhere else has a liquid setter and a volume getter. I think the one described here makes more sense because it gives a use for every variable

File: src/oop/special_class_method/hard/q13/QuestionText.md [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+- Methods:
+    - a `setLiquid` method which accepts a `String` and uses it to set the value of the **liquid** attribute
+    - a `getVolume` method which returns the value stored in the **volume** attribute
+    - These methods must be visible from other classes.
+  
+### Distractors
+
+    - public void setColour(String a)
+    - public void setVolume(double a)
+    - return this.year;
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q13/QuestionText.md:32]: this distractor seems a bit odd to me, since we normally don't have ones that are unrelated to the question

File: src/oop/special_class_method/hard/q13/QuestionText.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    - a `Cup` constructor that is visible to other classes and accepts a `String`, a `double` and another `String` and
+      uses it to initialize the **colour**, **volume**, and **liquid** attributes.
+
+- Methods:
+    - a `setLiquid` method which accepts a `String` and uses it to set the value of the **liquid** attribute
+    - a `getVolume` method which returns the value stored in the **volume** attribute
+    - These methods must be visible from other classes.
+  
+### Distractors
+
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q13/QuestionText.md:29]: could potentially also be distractors for attribute definition

File: src/oop/special_class_method/hard/q13/MainTest.java [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+    @BeforeEach
+    public void setUp() throws Throwable {
+        String classString = "oop.special_class_method.hard.q13." + className;
+        this.testClass = new ObjectTest(classString);
+    }
+
+    @Test
+    public void alienClassHasCorrectAttributes() {
+        assertTrue(testClass.hasField(attributeName1),
+                "Your " + className + " class is missing the " + attributeName1 + " attributes, or it is spelled incorrectly.");
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q13/MainTest.java:35]: change "attributes" to singular, here and in the other two error messages

**Review by SamamthaH — CHANGES_REQUESTED** [review]

File: src/oop/special_class_method/hard/q13/QuestionText.md (hunk 1 above)
Comment (KaiKhademi) [reply src/oop/special_class_method/hard/q13/QuestionText.md:7]: Its a parsons so I think for them to make 3 same type methods would be redundant and making one of each test their abilities better.

File: src/oop/special_class_method/hard/q13/MainTest.java [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+    private ObjectTest testClass;
+
+    @BeforeEach
+    public void setUp() throws Throwable {
+        String classString = "oop.special_class_method.hard.q13." + className;
+        this.testClass = new ObjectTest(classString);
+    }
+
+    @Test
+    public void alienClassHasCorrectAttributes() {
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q13/MainTest.java:33]:
```suggestion
    public void cupClassHasCorrectAttributes() {
```

File: src/oop/special_class_method/hard/q13/MainTest.java [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    @ParameterizedTest
+    @MethodSource("inputProvider")
+    public void correctGetVolumeMethod(String value1, double value2, String value3) throws Throwable {
+        Object[][] arguments = {
+                {value1, String.class},
+                {value2, double.class},
+                {value3, String.class}
+        };
+        Object testInstance = testClass.createInstance(arguments);
+        Object getMethodOutput = testClass.callMethod(getAttributeMethodName2, testInstance);
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q13/MainTest.java:101]: use .hasMethod first

File: src/oop/special_class_method/hard/q13/MainTest.java [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+        };
+        Object testInstance = testClass.createInstance(arguments);
+        Object[][] setMethodArguments = {
+                {updatedValue, String.class}
+        };
+        Class<?>[] methodModifierClasses = {
+                String.class
+        };
+        assertTrue(testClass.hasMethod(setAttributeMethodName3, methodModifierClasses, Void.TYPE),
+                "Your " + setAttributeMethodName3 + " method does not have the correct name, return type, or parameters.");
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q13/MainTest.java:124]: this assert should be split so return type has it's own error message

**Review by SamamthaH — CHANGES_REQUESTED** [review]

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]
See feedback on other questions in this category and apply it to this PR.

File: src/oop/special_class_method/hard/q13/QuestionText.md [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+### Cup Class:
+
+Your `Cup` class must have the following items:
+
+- Attributes:
+    - A `String` called **colour**
+    - A `double` called **volume**
+    - A `String` called **liquid**
+    - These are not visible from other classes
```
Comment (Hedgemon4) [inline src/oop/special_class_method/hard/q13/QuestionText.md:16]:
```suggestion
    - These attributes should not be visible from other classes.
```

File: src/oop/special_class_method/hard/q13/QuestionText.md [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+- Constructor:
+    - a `Cup` constructor that is visible to other classes and accepts a `String`, a `double` and another `String` and
+      uses it to initialize the **colour**, **volume**, and **liquid** attributes.
+
+- Methods:
+    - a `setLiquid` method which accepts a `String` and uses it to set the value of the **liquid** attribute. It must ]
+      not return anything.
```
Comment (Hedgemon4) [inline src/oop/special_class_method/hard/q13/QuestionText.md:24]:
```suggestion
    - a `setLiquid` method which accepts a `String` and uses it to set the value of the **liquid** attribute. It must not return anything.
```

File: src/oop/special_class_method/hard/q13/MainTest.java [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+    }
+
+
+    private static Stream<Arguments> inputProvider() {
+        return Stream.of(
+                Arguments.of("Red", 3423.4, "Milk", "Water"),
+                Arguments.of("Black", 54.5, "Water", "Lava"),
+                Arguments.of("Blue", 934.3, "Mercury", "Juice"),
+                Arguments.of("White", 49374.43, "Blood", "Water"),
+                Arguments.of("Yellow", 434.43, "Lava", "Whiskey")
```
Comment (Hedgemon4) [inline src/oop/special_class_method/hard/q13/MainTest.java:61]: Add a constructor with empty strings and 0, please.

File: src/oop/special_class_method/hard/q13/MainTest.java [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+        _assertEquals(value1, testClass.getFieldValue(classInstance, attributeName1),
+                "Your " + className + " constructor does not initialize the " + attributeName1 + " attribute to the correct value.");
+        _assertEquals(value2, testClass.getFieldValue(classInstance, attributeName2),
+                "Your " + className + " constructor does not initialize the " + attributeName2 + " attribute to the correct value.");
+        _assertEquals(value3, testClass.getFieldValue(classInstance, attributeName3),
+                "Your " + className + " constructor does not initialize the " + attributeName3 + " attribute to the correct value.");
+    }
+
+    @Test
+    public void getVolumeIsDefinedCorrectly() {
```
Comment (Hedgemon4) [inline src/oop/special_class_method/hard/q13/MainTest.java:93]: You can just remove this test as you cover it in the `correctGetVolumeMethod` test.

File: src/oop/special_class_method/hard/q13/MainTest.java [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+        assertTrue(testClass.hasModifier(getAttributeMethodName2, null, "public"), String.format(incorrectModifierMessage, getAttributeMethodName2, className));
+        assertTrue(testClass.hasReturnType(getAttributeMethodName2, null, double.class), String.format(incorrectReturnType, getAttributeMethodName2, className));
+        Object testInstance = testClass.createInstance(arguments);
+        Object getMethodOutput = testClass.callMethod(getAttributeMethodName2, testInstance);
+        _assertEquals(value2, getMethodOutput,
+                "Your " + getAttributeMethodName2 + " method does not return the value of the " + attributeName2 + " attribute.");
+    }
+
+    @Test
+    public void setLiquidIsDefinedCorrectly() {
```
Comment (Hedgemon4) [inline src/oop/special_class_method/hard/q13/MainTest.java:123]: Same thing with this test.

File: src/oop/special_class_method/hard/q13/MainTest.java (hunk 10 above)
Comment (Hedgemon4) [reply src/oop/special_class_method/hard/q13/MainTest.java:61]: You should also make it so the set method sets to an empty string from a non-empty one and vice versa.

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]
